packages/multichain/src/middlewares/multichainMethodCallValidator.ts

```diff
@@ -18,45 +13,92 @@ import { makeCustomResolver } from '@open-rpc/schema-utils-js/build/parse-open-r
 import type { Schema, ValidationError } from 'jsonschema';
 import { Validator } from 'jsonschema';
 
+export type MethodCallValidationSchemaError = {
+  message: string;
+  param: string;
+  path: (string | number)[];
+  schema: string | Schema;
+  got: unknown;
+};
+
+export type MethodCallValidatorNoSchemaError = {
+  message: string;
+  expected?: Json;
+  got: unknown;
+};
+
+export type MethodCallValidationError =
+  | MethodCallValidationSchemaError
+  | MethodCallValidatorNoSchemaError;
+
 const transformError = (
   error: ValidationError,
   param: ContentDescriptorObject,
   got: unknown,
-) => {
+): MethodCallValidationError => {
   // if there is a path, add it to the message
   const message = `${
     param.name + (error.path.length > 0 ? `.${error.path.join('.')}` : '')
   } ${error.message}`;
 
   return {
-    code: -32602, // TODO: could be a different error code or not wrapped in json-rpc error, since this will also be wrapped in a -32602 invalid params error
     message,
-    data: {
-      param: param.name,
-      path: error.path,
-      schema: error.schema,
-      got,
-    },
+    param: param.name,
+    path: error.path,
+    schema: error.schema,
+    got,
   };
 };
 
+const checkForInvalidParams = (
+  params: JsonRpcParams | undefined,
+  paramsToCheck: ContentDescriptorObject[],
+) => {
+  const errors: MethodCallValidationError[] = [];
+  const numRequiredParams = (paramsToCheck as ContentDescriptorObject[]).filter(
+    (p) => p.required,
+  ).length;
+
+  let paramsLength = 0;
+  if (Array.isArray(params)) {
+    paramsLength = params.length;
+  } else if (isObject(params)) {
+    paramsLength = Object.keys(params).length;
+  }
+
+  if (
+    (numRequiredParams > paramsLength && numRequiredParams > 0) ||
+    paramsLength > paramsToCheck.length
+  ) {
+    errors.push({
+      message: `Invalid number of parameters.`,
+      expected: paramsToCheck.length,
```

Comment:
does numRequiredParams need to go in here too?

Reply:
i needs to be `paramsToCheck.length` or `numRequiredParams` depending on which part of the `if` it hits. need to refactor it to 2 if statements to make it more clear and fix the bug.